ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala

```diff
@@ -215,23 +223,55 @@ class SubmissionValidator[LogResult](
               logger.error("Unexpected failure during submission validation.", exception)
               Success(Left(ValidationError(exception.getLocalizedMessage)))
           }
-        }
       case _ =>
         Future.successful(
           Left(ValidationError(s"Failed to parse submission, correlationId=$correlationId")))
     }
 
   private def flattenInputStates(
-      inputs: Map[DamlStateKey, Option[DamlStateValue]]): Map[DamlStateKey, DamlStateValue] =
+      inputs: Map[DamlStateKey, Option[DamlStateValue]]
+  ): Map[DamlStateKey, DamlStateValue] =
     inputs.collect {
       case (key, Some(value)) => key -> value
     }
 
-  object Metrics {
+  private final class TimedLedgerStateAccess(delegate: LedgerStateAccess[LogResult])
+      extends LedgerStateAccess[LogResult] {
+    override def inTransaction[T](
+        body: LedgerStateOperations[LogResult] => Future[T]
+    ): Future[T] = {
+      val acquireStopped = new AtomicBoolean(false)
+      val acquireTimer = Metrics.acquireTransaction.time()
+      delegate
+        .inTransaction { operations =>
+          if (acquireStopped.compareAndSet(false, true)) {
+            acquireTimer.stop()
+          }
+          body(operations)
+            .transform(result => Success((result, Metrics.releaseTransaction.time())))
+        }
+        .transform {
+          case Success((result, releaseTimer)) =>
+            releaseTimer.stop()
+            result
+          case Failure(exception) =>
+            if (acquireStopped.compareAndSet(false, true)) {
+              acquireTimer.stop()
+            }
+            Failure(exception)
+        }
+    }
+  }
+
+  private object Metrics {
     private val prefix = MetricRegistry.name("daml", "kvutils", "submission", "validator")
 
     val openEnvelope: Timer =
       metricRegistry.timer(MetricRegistry.name(prefix, "open_envelope"))
+    val acquireTransaction: Timer =
+      metricRegistry.timer(MetricRegistry.name(prefix, "acquire_transaction"))
```

Comment:
Shouldn't these be `acquire_transaction_lock` and `release_transaction_lock` instead? I.e., acquiring or releasing a transaction doesn't sound right.

Reply:
@miklos-da I think @SamirTalwar-DA wanted to call the metric in a way that is independent from the actual acquisition mechanism and I tend to agree.

Reply:
Done. 🙂